### PR TITLE
fix: throw proper error for not enough resources to cover fee

### PR DESCRIPTION
**Pull request by Torres-ssf**
- Close: https://github.com/FuelLabs/fuels-ts/issues/2127

### CODE REVIEW

File: packages/account/src/providers/provider.ts [hunk 1]
```diff
@@ -461,6 +461,15 @@ export default class Provider {
     const fetchFn = Provider.getFetchFn(this.options);
     const gqlClient = new GraphQLClient(this.url, {
       fetch: (url: string, requestInit: RequestInit) => fetchFn(url, requestInit, this.options),
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      responseMiddleware: ({ response }: any) => {
```
[the remark this answers is not in the document]
Comment (Torres-ssf) [reply packages/account/src/providers/provider.ts:465]: According to `graphql-request` types:
```ts
responseMiddleware: (response: Response<unknown> | Error) => {
```
 However, that is not the case. 

The parameter received by the middleware function has the `response` object within it along with other properties as well.

And it is never an Error instance also

File: packages/account/src/account.ts [hunk 2]
```diff
@@ -296,14 +291,39 @@ export class Account extends AbstractAccount {
...
+
+      if (totalBaseAssetOnInputs.gt(totalBaseAssetRequiredWithFee)) {
+        needsToBeFunded = false;
+      } else {
+        missingQuantities = [
+          {
+            amount: totalBaseAssetRequiredWithFee.sub(totalBaseAssetOnInputs),
+            assetId: BaseAssetId,
+          },
+        ];
```
[the remark this answers is not in the document]
Comment (Torres-ssf) [reply packages/account/src/account.ts:323]: No, once inside the loop, the `needsToBeFunded` only needs to be updated when:
```ts
totalBaseAssetOnInputs.gt(totalBaseAssetRequiredWithFee)
```

File: packages/fuel-gauge/src/funding-transaction.test.ts [hunk 3]
```diff
@@ -190,4 +192,58 @@ describe(__filename, () => {
...
+     * will fail after being submitted with Error "InsufficientInputAmount". This error is
+     * misleading because in reality the user has not enough coins to cover the fee. The query
+     * `getResourcesToSpend` is smart enough to return way more funds than the target amount. But since
+     * the user balance was only slightly above the target ( initially 1078 ), and because the user
+     * has its low balance spread in many UTXOs, the fee will be higher than the user balance after
+     * funding the TX. By trying to fund the TX again, we can force the proper error to be thrown.
+     */
+    await expectToThrowFuelError(
+      () => sender.fund(request, requiredQuantities, maxFee),
+      new FuelError(FuelError.CODES.INVALID_REQUEST, 'not enough coins to fit the target')
```
[the remark this answers is not in the document]
Comment (maschad) [reply packages/fuel-gauge/src/funding-transaction.test.ts:244]: Right, I should have made the comment inline with the client code, but I still think that the error type could be more descriptive.

File: packages/account/src/providers/provider.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Torres-ssf) [reply packages/account/src/providers/provider.ts]: https://github.com/FuelLabs/fuels-ts/pull/2128/commits/234db4ea1060568edac106e9e7e3afb9e849f3f0

File: packages/fuel-gauge/src/funding-transaction.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (arboleya) [reply packages/fuel-gauge/src/funding-transaction.test.ts:244]: Yes, I guess `TRANSACTION_FAILED` feels right.

[the remark this answers is not in the document]
Comment (Torres-ssf) [reply packages/fuel-gauge/src/funding-transaction.test.ts:230]: @arboleya @maschad I've created an issue for this: https://github.com/FuelLabs/fuels-ts/issues/2208

[the remark this answers is not in the document]
Comment (arboleya) [reply packages/fuel-gauge/src/funding-transaction.test.ts:230]: IIUC @maschad's review, the final request here would be only to rename the error code as `TRANSACTION_FAILED`, which may be more fitting than `INVALID_REQUEST`.

[the remark this answers is not in the document]
Comment (maschad) [reply packages/fuel-gauge/src/funding-transaction.test.ts:230]: Thanks for the clarification @Torres-ssf that makes sense

**Review by maschad — APPROVED** [review]
Great work!

**Pull request merged**